internal/sql/pg/repl_changeset.go

```diff
@@ -1,29 +1,253 @@
+	var columnStr, placeholderStr string
+	if len(rel.Columns) > 0 {
+		columnStr = rel.Columns[0].Name
+		placeholderStr = "$1"
+		for i := 1; i < len(rel.Columns); i++ {
+			columnStr += ", " + rel.Columns[i].Name
+			placeholderStr += ", $" + fmt.Sprint(i+1)
+		}
+	}
+
+	// Conflict resolution: DO NOTHING
+	// Any conflicts will be ignored, in favor of whatever already exists on the new network.
+	insertSql := fmt.Sprintf("INSERT INTO %s.%s (%s) VALUES (%s) ON CONFLICT DO NOTHING", rel.Schema, rel.Table, columnStr, placeholderStr)
+
+	_, newVals, err := c.DecodeTuples(rel)
+	if err != nil {
+		return err
+	}
+
+	_, err = tx.Execute(ctx, insertSql, newVals...)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// applyUpdates applies all updates in the changeset to the database.
+// Apply updates only if the oldValues in the old network are same as the current record in the new network.
+// If not, discard the update in favor of whatever data exists on the new network
+func (c *ChangesetEntry) applyUpdates(ctx context.Context, tx sql.DB, rel *Relation) error {
+	if len(c.OldTuple) != len(c.NewTuple) {
+		return fmt.Errorf("old and new tuples have different lengths")
+	}
+
+	updateSql := fmt.Sprintf("UPDATE %s.%s SET ", rel.Schema, rel.Table)
+	for i, col := range rel.Columns {
+		if i > 0 {
+			updateSql += ", "
+		}
+		updateSql += col.Name + " = $" + fmt.Sprint(i+1)
+	}
+
+	// Conflict resolution:
+	// If new network's current record is same as the oldValues in the old network, then update the record
+	// Else, discard the update in favor of whatever data exists on the new network
+	oldVals, newVals, err := c.DecodeTuples(rel)
+	if err != nil {
+		return err
+	}
+
+	var oldArgs []any
+	cnt := 1
+	whereClause := ""
+	for i, v := range oldVals {
+		if i > 0 {
+			whereClause += " AND "
+		}
+
+		if v == nil {
+			whereClause += fmt.Sprintf("%s IS NULL", rel.Columns[i].Name)
+		} else {
+			whereClause += fmt.Sprintf("%s = $%d", rel.Columns[i].Name, cnt+len(newVals))
+			oldArgs = append(oldArgs, v)
+			cnt++
+		}
+	}
+
+	_, err = tx.Execute(ctx, updateSql+" WHERE "+whereClause, append(newVals, oldArgs...)...)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// applyDeletes applies all deletes in the changeset to the database.
+// If the record in the new network is same as the oldValues in the old network, then delete the record
+// Else, discard the delete in favor of whatever data exists on the new network
+func (ce *ChangesetEntry) applyDeletes(ctx context.Context, tx sql.DB, rel *Relation) error {
+	deleteSql := fmt.Sprintf("DELETE FROM %s.%s WHERE ", rel.Schema, rel.Table)
+
+	record, _, err := ce.DecodeTuples(rel)
+	if err != nil {
+		return err
+	}
+
+	whereClause := ""
+	var args []any
+	cnt := 1
+	for i, v := range record {
+		if i > 0 {
+			whereClause += " AND "
+		}
+		if v == nil {
+			whereClause += fmt.Sprintf("%s IS NULL", rel.Columns[i].Name)
```

Comment:
Normally this suggestion would be a micro optimization, but these methods are likely to be processing a significant amount of data: Prefer `var whereClause strings.Builder` and do `whereClause.WriteString(" AND ")` and `fmt.Fprintf(&whereClause, "%s IS NULL", rel.Columns[i].Name)` etc.